### ISSUE TITLE
 Flutter can load old or stale application code, randomly.

### ISSUE THREAD

**Issue by esDotDev** [flutter/flutter#72868]
**Problem**
On android (and maybe elsewhere), Flutter can load old or stale application code, randomly. This leads to extremely frustrating test and debug cycles.

I have been experiencing this with Flutter, for over a year (since I started working with it), but it's always been hard to catch, and usually related to some runtime errors. It's my top issue with productivity in Flutter, but I've never had clean reproduction steps, it just happens, randomly, and usually at the most in-opportune times. 

Today I think I've found a reproduceable way to make it happen when testing the Restoration API. And hopefully this can lead to fixing the issue overall. 

https://user-images.githubusercontent.com/736973/102984559-005bd400-44cb-11eb-8631-5d711129d304.mp4

**@29s, 39s, and 50s in this video, you can see Flutter loading the previous version of the application 3 times, even though I have repeatedly hot-reloaded a new view.** The text has clearly been set to end-alignment, but somehow it reverts to center. 

When I hot reload a 2d time, the new view comes back. When the app is restored, the old view comes back!

This is the stuff of developer nightmare fuel. Somewhere is a cached version of the app, that should definitely not be there.

**Please, whatever you have to do, make sure that we never get an old version of the app running.**. It might seem like a small issue, but at production, at scale, when you're chasing bugs, and you can never be quite sure if your new code is actually being ran... it's really tough to work with, and mentally draining.

It's exacerbated by the frameworks tendency to throw a ton of assert errors, which generally seem to precede this behavior. This makes it hard to spot on small projects as well, it tends to manifest when you're at scale, and when you're in the middle of crunch time. When you do a full stop and deploy, it goes away, which makes it _extremely_ hard to catch.

```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel dev, 1.26.0-1.0.pre, on Microsoft Windows [Version 10.0.19041.630], locale en-US)
[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[√] Chrome - develop for the web
[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.7.6)
[√] Android Studio (version 4.0)
[√] VS Code (version 1.52.1)
[√] Connected device (3 available)
```

Repro Steps:
1. In Android Developer settings, set to "Don't Keep Activities"
2. Install and run the following code
3. Make changes to the _RestorableCounterState, and force the app to be reloaded, and observe as it easily gets out of sync.
```dart
import 'package:flutter/material.dart';

void main() => runApp(RestorationExampleApp());

class RestorationExampleApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      restorationScopeId: 'app',
      title: 'Restorable Counter',
      home: RestorableCounter(restorationId: 'counter'),
    );
  }
}

class RestorableCounter extends StatefulWidget {
  RestorableCounter({Key key, this.restorationId}) : super(key: key);

  final String restorationId;

  @override
  _RestorableCounterState createState() => _RestorableCounterState();
}

class _RestorableCounterState extends State<RestorableCounter> with RestorationMixin {
  RestorableInt _counter = RestorableInt(0);

  @override
  String get restorationId => widget.restorationId;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.end,
          children: <Widget>[
            Text('You have pushed the button this many times:'),
            Text('${_counter.value}', style: Theme.of(context).textTheme.headline4),
            SizedBox(height: 100)
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => setState(() => _counter.value++),
        child: Icon(Icons.add),
      ),
    );
  }

  @override
  void restoreState(RestorationBucket oldBucket, bool initialRestore) => registerForRestoration(_counter, "_counter");
}
```

**Comment by TahaTesser** [comment]
Hi @esDotDev 
Amazing attention to detail, Thanks for filing the issue,  I can reproduce on multiple channels except `stable`, code sample doesn't seem compatible for `stable`, `restorationScopeId` isn't implemented on this channel. I can reproduce the behavior from `beta` to `master`

<details> 
<summary>multiple flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel beta, 1.25.0-8.1.pre, on macOS 11.1 20C69 darwin-x64,
    locale en-GB)
    • Flutter version 1.25.0-8.1.pre at /Users/tahatesser/Code/flutter_beta
    • Framework revision 8f89f6505b (8 days ago), 2020-12-15 15:07:52 -0800
    • Engine revision 92ae191c17
    • Dart version 2.12.0 (build 2.12.0-133.2.beta)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[!] Android Studio
    • Android Studio at /Applications/Android/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Android Studio not found at /Applications/Android/Contents
    • Try updating or re-installing Android Studio.

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.17.0

[✓] Connected device (3 available)
    • RMX2001 (mobile)     • EUYTFEUSQSRGDA6D          • android-arm64  •
      Android 10 (API 29)
    • Taha’s iPad (mobile) • 00008020-000255113EE8402E • ios            • iOS
      14.3
    • Chrome (web)         • chrome                    • web-javascript • Google
      Chrome 87.0.4280.88

! Doctor found issues in 1 category.
```

```bash
[✓] Flutter (Channel dev, 1.26.0-1.0.pre, on macOS 11.1 20C69 darwin-x64, locale
    en-GB)
    • Flutter version 1.26.0-1.0.pre at /Users/tahatesser/Code/flutter_dev
    • Framework revision 63062a6443 (10 days ago), 2020-12-13 23:19:13 +0800
    • Engine revision 4797b06652
    • Dart version 2.12.0 (build 2.12.0-141.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[!] Android Studio
    • Android Studio at /Applications/Android/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Android Studio not found at /Applications/Android/Contents
    • Try updating or re-installing Android Studio.

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.17.0

[✓] Connected device (4 available)
    • RMX2001 (mobile)     • EUYTFEUSQSRGDA6D          • android-arm64  •
      Android 10 (API 29)
    • Taha’s iPad (mobile) • 00008020-000255113EE8402E • ios            • iOS
      14.3
    • macOS (desktop)      • macos                     • darwin-x64     • macOS
      11.1 20C69 darwin-x64
    • Chrome (web)         • chrome                    • web-javascript • Google
      Chrome 87.0.4280.88

! Doctor found issues in 1 category.
```

```bash
[✓] Flutter (Channel master, 1.26.0-2.0.pre.117, on macOS 11.1 20C69 darwin-x64,
    locale en-GB)
    • Flutter version 1.26.0-2.0.pre.117 at
      /Users/tahatesser/Code/flutter_master
    • Framework revision 15c009b805 (10 hours ago), 2020-12-22 20:09:03 -0500
    • Engine revision b7cd3f8e70
    • Dart version 2.12.0 (build 2.12.0-173.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[!] Android Studio
    • Android Studio at /Applications/Android/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Android Studio not found at /Applications/Android/Contents
    • Try updating or re-installing Android Studio.

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.17.0

[✓] Connected device (4 available)
    • RMX2001 (mobile)     • EUYTFEUSQSRGDA6D          • android-arm64  •
      Android 10 (API 29)
    • Taha’s iPad (mobile) • 00008020-000255113EE8402E • ios            • iOS
      14.3
    • macOS (desktop)      • macos                     • darwin-x64     • macOS
      11.1 20C69 darwin-x64
    • Chrome (web)         • chrome                    • web-javascript • Google
      Chrome 87.0.4280.88

! Doctor found issues in 1 category
```

</details>

**Comment by Vulfpeck** [comment]
AFAIK when you do a debug run the only code that's persisted is from the initial launch and any subsequent hot reloads are lost when the app restarts which is what might be happening here.

**Comment by esDotDev** [comment]
Interesting. So perhaps there are other conditions where the app restarts, or loses some race condition, and the old app is presented? I find it will happen very often, where there was some errors, and then they are fixed, yet the old version with the errors just stubbornly continues to hot-reload in, over and over.

I would happily 10x my compile/deploy times if it meant a 100% chance I would get my latest code. On some projects, I've setup scripts to do a clean + run every single time, because it's just a total nightmare when it keeps popping up old code.

**Comment by Vulfpeck** [comment]
> So perhaps there are other conditions where the app restarts, or loses some race condition, and the old app is presented?

I'm not sure I understand this bit, but from my experience if you start a debug run -> make changes to app -> stop the debug run -> open the 'debug'  app installed you'll see the app in the state that you starting your debug session with. ie. all your hot-reload changes after the start will be long. 

This happens 100% of the time.

**Comment by esDotDev** [comment]
What I experience, is while working, I will hot reload, or hot restart, an old an version of the app will appear. When this happens you can reload many times, it won't help. Only a full stop, and restart can break out of this loop.  It wouldn't be as bad if you knew it had happened, but unless you have some strong visual cue like a color change, it can be very easy to miss.

**Comment by nt4f04uNd** [comment]
> What I experience, is while working, I will hot reload, or hot restart, an old an version of the app will appear. When this happens you can reload many times, it won't help. Only a full stop, and restart can break out of this loop. It wouldn't be as bad if you knew it had happened, but unless you have some strong visual cue like a color change, it can be very easy to miss.

for the record, i've never exprienced anything like this
i'm developing under windows, so that's highly likely might be macos related (sometimes i run debugger on vm macos in simulator, but i have really never met this). yep, sometimes debugger crashes, hags on forever or just doesn't want to reload, but nothing like @esDotDev described in this quote

<details>
<summary>flutter doctor -v</summary>
[✓] Flutter (Channel master, 1.24.0-8.0.pre.374, on Microsoft Windows [Version 10.0.19041.685], locale ru-RU)
    • Flutter version 1.24.0-8.0.pre.374 at c:\dev\src\flutter
    • Framework revision 183f0e797a (4 weeks ago), 2020-11-26 19:12:28 +0100
    • Engine revision 20caf54969
    • Dart version 2.12.0 (build 2.12.0-76.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at C:\Users\danya\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[✓] Android Studio (version 4.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 50.0.1
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[✓] VS Code (version 1.52.1)
    • VS Code at C:\Users\danya\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.18.0

[!] Connected device
    ! No devices available

! Doctor found issues in 1 category.
</details>

**Comment by justinmc** [comment]
CC @goderbauer

**Comment by kf6gpe** [comment]
Tentatively assigning this as a P4 --- a bug we'd like to fix time permitting, but not actively working on.

**Comment by canbrkkyGit** [comment]
Any update ? Same problem is still existing on Flutter 3.0.5

**Comment by exaby73** [comment]
I can no longer reproduce this issue on `Master (3.10.0-7.0.pre.46)`. Therefore, I will close this issue. If you can still reproduce this issue, please provide a code sample along with detailed steps to reproduce, and the output of `flutter doctor -v`. Thank you

<details>
<summary>Code Sample migrated to null safety</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const RestorationExampleApp());

class RestorationExampleApp extends StatelessWidget {
  const RestorationExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      restorationScopeId: 'app',
      title: 'Restorable Counter',
      home: RestorableCounter(restorationId: 'counter'),
    );
  }
}

class RestorableCounter extends StatefulWidget {
  const RestorableCounter({super.key, required this.restorationId});

  final String restorationId;

  @override
  State<RestorableCounter> createState() => _RestorableCounterState();
}

class _RestorableCounterState extends State<RestorableCounter>
    with RestorationMixin {
  final RestorableInt _counter = RestorableInt(0);

  @override
  String get restorationId => widget.restorationId;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text('You have pushed the button this many times:'),
            Text(
              '${_counter.value}',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
            const SizedBox(height: 100)
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => setState(() => _counter.value++),
        child: const Icon(Icons.add),
      ),
    );
  }

  @override
  void restoreState(RestorationBucket? oldBucket, bool initialRestore) =>
      registerForRestoration(_counter, "_counter");
}
```

</details>

<details><summary><code>flutter doctor -v</code> (Master)</summary>

```
[!] Flutter (Channel master, 3.10.0-7.0.pre.46, on macOS 13.3.1 22E261 darwin-arm64, locale en-US)
    • Flutter version 3.10.0-7.0.pre.46 on channel master at /Users/nabeelparkar/fvm/versions/master
    ! Warning: `dart` on your path resolves to /opt/homebrew/Cellar/dart/2.19.5/libexec/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nabeelparkar/fvm/versions/master. Consider adding /Users/nabeelparkar/fvm/versions/master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 59230759be (22 minutes ago), 2023-04-19 02:39:50 -0400
    • Engine revision 609f9d5364
    • Dart version 3.1.0 (build 3.1.0-15.0.dev)
    • DevTools version 2.23.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at /Users/nabeelparkar/Library/Android/sdk/
    • Platform android-33, build-tools 33.0.2
    • ANDROID_SDK_ROOT = /Users/nabeelparkar/Library/Android/sdk/
    • Java binary at: /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9862592/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.12.0

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /Applications/Brave Browser.app/Contents/MacOS/Brave Browser

[✓] Android Studio (version 2022.1)
    • Android Studio at /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/221.6008.13.2211.9619390/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9862592/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.77.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.62.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.3.1 22E261 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Brave Browser 112.1.50.119

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>
